### PR TITLE
Add Reducescatter op (NCCL, MPI, Gloo)

### PR DESCRIPTION
**Pull request by maxhgerlach**
## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs?
- [x] Did you write any tests to validate this change?  
- [x] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

This PR is an update and extension of @jessebenson's PR #1496 from late 2019 / early 2020.

It adds a new collective operation to Horovod, `hvd.reducescatter()` as illustrated by this figure from NVIDIA's documentation:
image

Since a gradient allreduce can be understood as a reducescatter of the gradients followed by an allgather, having both intermediate operations available provides Horovod users with extra flexibility in their Python training code. For instance, they might experiment with distributing their optimizer states over the worker instances to achieve one of the ZeRO schemes demonstrated in Microsoft's "DeepSpeed".

Included implementations:
- MPI and MPI GPU: mostly taken from Jesse's commits
- NCCL: Because there is no "nccReduceScatter**V**()", this required some extra attention for the case that a tensor cannot be divided over the Horovod processes without a remainder. In this case the first `tensor_shape.dim_size(0) % process_set_size` ranks receive a slightly larger portion of the reduced tensor than the later ranks (one extra slice each). I've implemented this case via an `ncclGroup` of `ncclReduce()` calls, similarly to how Horovod's `NCCLAllgather` simulates "ncclAllGather**V**" with a group of `ncclBroadcast()` calls.
- Gloo: For `reducescatter` there somehow is no high-level API like `gloo::allreduce()`, `gloo::allgatherv`, etc. However, there is an algorithm `gloo::ReduceScatterHalvingDoubling` which can be called directly. Edit: Sadly, this one does not work on macOS, where we need to use the `libuv` transport, rather than `tcp`.

oneCCL is not supported yet.

Bindings and tests are provided for TensorFlow, PyTorch, and MXNet.

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

**Comment by maxhgerlach** [comment]
Refreshed this to master

**Comment by maxhgerlach** [comment]
I think only one of the flaky tests failed in 1 of 20 runs. Not sure why this would fail. the build.

test_max_np (test.single.test_ray_elastic_v2)

File: horovod/common/ops/nccl_operations.cc [hunk 1]
```diff
@@ -836,5 +836,128 @@ Status NCCLAlltoall::Execute(std::vector<TensorTableEntry>& entries,
...
+                   process_set_rank * recvcounts[0] * element_size;
+    // ncclReduceScatter() will be performed in place on the fusion buffer, cf.:
+    // https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/inplace.html
+  } else {
+    fused_input_data = first_entry.tensor->data();
+    buffer_data = (void*)first_entry.output->data();
+    recv_pointer = buffer_data;
+  }
+
+  bool same_output_shape = (recvcounts.front() == recvcounts.back());
```
[the remark this answers is not in the document]
Comment (maxhgerlach) [reply horovod/common/ops/nccl_operations.cc:886]: Hi @romerojosh, thank you so much for your first-pass review, and sorry for taking so long to respond! I was a bit caught up.

I've added a clarifying comment here. Also I agree that scattering the data more evenly as you suggest would make sense. In my current performance-sensitive application of this op I always make sure that the number of ranks evenly divides the sizes of all involved tensors, so I haven't given this much thought so far and haven't changed this part of the code from the original PR.

I'll see if I can change distribution pattern next.

File: horovod/mxnet/mpi_ops.py [hunk 2]
```diff
@@ -449,3 +454,59 @@ def alltoall(tensor, splits=None, name=None, priority=0, process_set=global_proc
         return output, output_received_splits
     else:
         return output
+
+def reducescatter(tensor, op=Average, name=None, priority=0,
```
[the remark this answers is not in the document]
Comment (maxhgerlach) [reply horovod/mxnet/mpi_ops.py:491]: Good idea! I've added the `op` API to the MXNet `allreduce` functions and deprecated the `average` argument.

Also updated the related deprecation notices for TF and Torch. If we are going to do another minor preliminary release, Horovod 1.0 might then be a good point to drop the `average` argument everywhere.

**Review by romerojosh — APPROVED** [review]
Thanks a lot for making the suggested changes @maxhgerlach and excellent work on this PR. LGTM!

**Comment by maxhgerlach** [comment]
Cheers, I'm stoked to get this one in!

I've merged master once more to make sure no framework update from #3426 breaks any tests. Will land once the CI passes.
